Envoy rebase 2017-11-30

jrajahalme opened the pull request:
Rebase Envoy to current (as of 11/30/2017) Envoy upstream master. Fixes several crashing bugs.

Upstream Envoy has changed to always build with the latest stable Bazel release. As of now it is 0.8.0. Since we depend on a specific commit on Envoy upstream, we must use the Bazel version that was stable at the time of that commit was merged. Hence this PR changes the Vagrantfile to pull Bazel version 0.8.0.

envoy/Makefile adds new targets that can be used for archiving and restoring pre-built Bazel dependencies (including Envoy itself).

Envoy v2 APIs are still work-in-progress and have been changed in ways that required changes also on the cilium side.

Envoy has instituted a naming convention for well-known filters in form of a "envoy." prefix. Json (v1) configs are compatible without changes, but gRPC (v2) configs must use the new naming convention. Cilium filters have been renamed in the same style with a "cilium." prefix.

raybejjani commented on Vagrantfile:
```diff
@@ -31,6 +31,10 @@ chown -R vagrant:vagrant $GOPATH
+echo "deb [arch=amd64] http://storage.googleapis.com/bazel-apt stable jdk1.8" | tee /etc/apt/sources.list.d/bazel.list
+curl https://bazel.build/bazel-release.pub.gpg | apt-key add -
+apt-get update
+apt-get install -y bazel=0.8.0
```
Perhaps move this section to the start of the script? the apt-get update we do there would pull in the new apt repository, and you could include the bazel install on the already existing line (or keep it separate if you want to make it more obvious).
In any case, the other `apt-get update` seems to `|| true` (I imagine it exits non-0 on a noop or something) so we should do that here too.

raybejjani commented on envoy/Makefile:
```diff
@@ -53,69 +55,98 @@ CILIUM_GOS = $(CILIUM_PROTOS:.proto=.pb.go)
+bazel-restore: $(BAZEL_ARCHIVE)
+	echo "Clearing & restoring bazel archive ($(BAZEL_ARCHIVE))..."
+	-sudo rm -Rf $(BAZEL_CACHE)
+	-mkdir $(dir $(BAZEL_CACHE))
```
`mkdir -p` ? Since BASEL_CACHE can be set by the user it might end up being on a path that needs intermediate directories created.

raybejjani commented on envoy/accesslog.cc:
```diff
@@ -54,18 +54,18 @@ AccessLog::AccessLog(std::string path) : path_(path), fd_(-1), open_count_(1) {}
+void AccessLog::Entry::InitFromRequest(
+    std::string listener_id, const Network::Connection *conn,
+    const Http::HeaderMap &headers, const Envoy::AccessLog::RequestInfo &info,
+    const Router::RouteEntry *route) {
```
These were aligned before, and easier to read. I'm not sure if it was intentional to change the formatting, if so, ignore me!

joestringer commented on Makefile:
```diff
@@ -1,13 +1,15 @@
 GOLANGVERSION = $(shell go version 2>/dev/null | grep -Eo '(go[0-9].[0-9])')
-GOLANG_SRCFILES=$(shell for pkg in $GOFILES; do find $(pkg) -name *.go -print; done | grep -v /vendor/)
-BPF_SRCFILES=$(shell find bpf/ -name *.[ch] -print)
+GOLANG_SRCFILES=$(shell for pkg in $(subst github.com/cilium/cilium/,,$(GOFILES)); do find $$pkg -name *.go -print; done | grep -v /vendor/)
```
This excludes `vendor/` at the end of the line, but shouldn't vendor already be excluded from the line in GOFILES above?

joestringer commented on Makefile:
```diff
@@ -106,30 +108,36 @@ tags: $(GOLANG_SRCFILES) $(BPF_SRCFILES)
+
+docker-image-runtime:
+	cd contrib/packaging/docker && docker build -t "cilium/cilium-runtime:$(UTC_DATE)" -f Dockerfile.runtime .
+	echo "Update Dockerfile with the new tag and push like this when ready:\ndocker push cilium/cilium-runtime:$(UTC_DATE)"
```
This description differs a little from the one under `docker-image`; is the "update dockerfile with the new tag" a separate step that requires input from the developer?

joestringer commented on Makefile:
```diff
@@ -106,30 +108,36 @@ tags: $(GOLANG_SRCFILES) $(BPF_SRCFILES)
+	echo "Push like this when ready:\ndocker push cilium/cilium:$(DOCKER_IMAGE_TAG)"
+
+docker-image-runtime:
+	cd contrib/packaging/docker && docker build -t "cilium/cilium-runtime:$(UTC_DATE)" -f Dockerfile.runtime .
```
Seems like previously we also passed `-q` here and to other `docker build` commands; Were you intending to drop this, or just removed it for debugging purposes while developing this series?

joestringer commented on tests/k8s/cluster/cluster-manager.bash:
```diff
@@ -185,7 +185,12 @@ function install_kubeadm_dependencies(){
+    curl -s https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
+    sudo bash -c "cat <<EOF > /etc/apt/sources.list.d/docker.list
+deb [arch=amd64] https://download.docker.com/linux/ubuntu xenial stable
+EOF
```
Should these steps be in the image we build rather than the test scripts here?

joestringer replied:
Should these steps be in the image we build rather than the tset scripts here?

joestringer commented on Vagrantfile:
```diff
@@ -31,6 +31,11 @@ chown -R vagrant:vagrant $GOPATH
+chmod +x bazel-0.8.1-installer-linux-x86_64.sh
+sudo -E ./bazel-0.8.1-installer-linux-x86_64.sh
+sudo -E mv /usr/local/bin/bazel /usr/bin
+rm bazel-0.8.1-installer-linux-x86_64.sh
```
Might be a little easier to update in future if each of these `0.8.1` strings refer to a ${BAZEL_VERSION} variable.

joestringer reviewed (approved):
Minor comments inline. Nothing blocking.

ianvernon reviewed (approved):
Vendor changes look fine. Vagrant changes have been removed from this PR, but GitHub doesn't remove CODEOWNERS after the changes to code mapped to specific teams are removed.

jrajahalme committed:
vendor: Add github.com/lyft/protoc-gen-validate

This is used by pkg/envoy/api in the following patch.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Upstream rebase.

Update Envoy reference to upstream Envoy master as of 2017-11-30 and
update cilium filters accordingly.

While upstream Envoy has moved to always building with the latest
stable bazel, we must build with the version that was used for the
upstream Envoy branch we depend on. Currently this means Bazel 0.8.1.

Bazel 0.8.0 changed the naming of directories in bazel-out symlink,
e.g., from "local-fastbuild" to "k8-fastbuild". "k8" here is the
default CPU target for Linux, which used to be "local"
before. Apparently this prefix will be different for different
targets, so the makefile should not assume to know them.

The reason for using those paths was that the "bazel-bin" symlink may
be changed when running bazel. If two targets are built (e.g.,
"release" and "debug") these will produce two different binaries, but
the "bazel-bin" will only lead to last of them (in the build order).

Bazels build directories are also dependent on the current user, so
they will be different for "vagrant" and "root" users. This is why
"install" target, that needs to be run as root, should not run bazel.

This patch solves this dilemma by simplifying the Makefile to only
ever produce a single envoy binary.

Makefile is also modified to explicitly check that the build targets
are not run as root, and that the install target is run as root.

Finally, support for archiving the bazel cache is added:

$ make bazel-archive - Build Envoy and tests using the current bazel
                       build cache, remove the Envoy and test binaries
                       and archive the bazel archive to
                       ~/bazel-cache.tar.bz2.

$ make bazel-clean-archive - Clear the current cache, build Envoy and
                             tests from scratch, remove the Envoy and
                             test binaries and archive the resulting
                             bazel archive to ~/bazel-cache.tar.bz2.

$ make bazel-restore - Clear the current cache and restore it from
                       ~/bazel-cache.tar.bz2.

After 'bazel-restore' the build (e.g., make tests) should only link
Envoy and test binaries. 'bazel-archive' can result in a bigger
archive, as it can carry on files that are strictly not needed any
more.

Envoronment variable BAZEL_ARCHIVE can be set to override the default
archive file path.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
